File: sample-app-java/app/src/main/java/com/bluejeans/android/sdksample/isc/usecases/RemoteLearningFragment.java
package com.bluejeans.android.sdksample.isc.usecases;

import android.content.res.Configuration;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.TextureView;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.constraintlayout.widget.ConstraintSet;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import com.bjnclientcore.media.individualstream.StreamPriority;
import com.bjnclientcore.media.individualstream.StreamQuality;
import com.bjnclientcore.media.individualstream.VideoStreamConfiguration;
import com.bjnclientcore.media.individualstream.VideoStreamConfigurationResult;
import com.bluejeans.android.sdksample.databinding.FragmentRemoteLearningBinding;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class RemoteLearningFragment extends Fragment {
    private static final String TAG = "RemoteLearningFragment";
    private FragmentRemoteLearningBinding binding = null;

    private RemoteLearningParticipant moderator = null;
    private List<RemoteLearningParticipant> students = new ArrayList<>();
    private List<TextureView> studentTextureViews = new ArrayList<>();

    private RemoteLearningViewModel viewModel;

    public RemoteLearningFragment() {
        // Required empty public constructor
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = FragmentRemoteLearningBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        studentTextureViews.add(0, binding.studentOne.participantTextureView);
        studentTextureViews.add(1, binding.studentTwo.participantTextureView);
        studentTextureViews.add(2, binding.studentThree.participantTextureView);
        studentTextureViews.add(3, binding.studentFour.participantTextureView);
        studentTextureViews.add(4, binding.studentFive.participantTextureView);
        studentTextureViews.add(5, binding.studentSix.participantTextureView);

        viewModel = new ViewModelProvider(this).get(RemoteLearningViewModel.class);
        viewModel.participantsObservable.observe(getViewLifecycleOwner(), studentsList -> {
            for (TextureView v : studentTextureViews) {
                v.setVisibility(View.GONE);
            }
            moderator = null;
            students.clear();
            students.addAll(studentsList);

            List<VideoStreamConfiguration> configurationList = new ArrayList<>();
            students.forEach(p -> {
                if (p.getParticipantId() != null) {
                    if (p.isModerator()) {
                        moderator = p;
                        configurationList.add(
                                new VideoStreamConfiguration(
                                        p.getParticipantId(),
                                        StreamQuality.R720p_30fps,
                                        StreamPriority.High
                                )
                        );
                    } else {
                        configurationList.add(
                                new VideoStreamConfiguration(
                                        p.getParticipantId(),
                                        StreamQuality.R90p_15fps,
                                        StreamPriority.Low
                                )
                        );
                    }
                }
            });

            VideoStreamConfigurationResult result = viewModel.setVideoConfiguration(configurationList);
            if (result instanceof VideoStreamConfigurationResult.Success) {
                updateModeratorUI();
                updateStudentsUI();
            } else {
                Log.e(TAG, "Failed to set stream configuration: " + result);
            }
        });
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (moderator != null) {
            viewModel.detachParticipantFromView(moderator.getParticipantId());
        }

        students.forEach(student -> {
            viewModel.detachParticipantFromView(student.getParticipantId());
        });
    }

    @Override
    public void onConfigurationChanged(@NonNull Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        if (newConfig.orientation == Configuration.ORIENTATION_PORTRAIT) {
            configurePortraitView();
        } else configureLandscapeView();
    }

    private void updateModeratorUI() {
        if (moderator != null) {
            binding.layoutMainStage.tvParticipantName.setText(moderator.getName());

            if (!moderator.isVideo()) {
                Log.i(TAG, "Detaching moderator video");
                viewModel.detachParticipantFromView(moderator.getParticipantId());

                binding.layoutMainStage.participantTextureView.setVisibility(View.GONE);
                binding.layoutMainStage.cnhAudioOnly.setVisibility(View.VISIBLE);
                binding.layoutMainStage.cnhAudioOnly.setText(getInitials(moderator.getName()));
            } else {
                binding.layoutMainStage.participantTextureView.setVisibility(View.VISIBLE);
                binding.layoutMainStage.cnhAudioOnly.setVisibility(View.GONE);
                viewModel.attachParticipantToView(moderator.getParticipantId(), binding.layoutMainStage.participantTextureView);
                setModeratorTextureViewConstraints();
            }
        } else {
            binding.layoutMainStage.tvParticipantName.setText("");
            binding.layoutMainStage.cnhAudioOnly.setVisibility(View.GONE);
        }
    }

    private String getInitials(String name) {

        String[] splitName = name.split(" ");
        if (splitName.length == 0) {
            return "";
        }
        if (splitName.length == 1) {
            return splitName[0].substring(0, 0);
        }
        StringBuilder finalName = new StringBuilder();
        int length = Math.min(splitName.length, 3);
        for (int i = 0; i < length; i++) {
            finalName.append(splitName[i].substring(0, 0));
        }
        return finalName.toString();
    }

    private void setModeratorTextureViewConstraints() {
        ConstraintSet set = new ConstraintSet();
        TextureView textureView = binding.layoutMainStage.participantTextureView;
        set.clone(binding.layoutMainStage.getRoot());
        set.connect(textureView.getId(), ConstraintSet.RIGHT, ConstraintSet.PARENT_ID, ConstraintSet.RIGHT);
        set.connect(textureView.getId(), ConstraintSet.LEFT, ConstraintSet.PARENT_ID, ConstraintSet.LEFT);
        set.connect(textureView.getId(), ConstraintSet.BOTTOM, ConstraintSet.PARENT_ID, ConstraintSet.BOTTOM);
        set.connect(textureView.getId(), ConstraintSet.TOP, ConstraintSet.PARENT_ID, ConstraintSet.TOP);

        if (moderator != null) {
            if (moderator.getWidth() > 0 && moderator.getHeight() > 0) {
                float ratio = ((float) moderator.getWidth() / moderator.getHeight());
                Log.i(TAG, "Ratio: " + ratio);
                set.setDimensionRatio(textureView.getId(), String.valueOf(ratio));
            }
        }

        set.applyTo(binding.layoutMainStage.getRoot());
    }

    private void setStudentTextureViewConstraints(int videoWidth, int videoHeight, TextureView textureView) {
        ConstraintSet set = new ConstraintSet();
        set.clone(binding.layoutMainStage.getRoot());
        set.connect(textureView.getId(), ConstraintSet.RIGHT, ConstraintSet.PARENT_ID, ConstraintSet.RIGHT);
        set.connect(textureView.getId(), ConstraintSet.LEFT, ConstraintSet.PARENT_ID, ConstraintSet.LEFT);
        set.connect(textureView.getId(), ConstraintSet.BOTTOM, ConstraintSet.PARENT_ID, ConstraintSet.BOTTOM);
        set.connect(textureView.getId(), ConstraintSet.TOP, ConstraintSet.PARENT_ID, ConstraintSet.TOP);

        if (moderator != null) {
            if (videoWidth > 0 && videoHeight > 0) {
                float ratio = ((float) videoWidth / videoHeight);
                Log.i(TAG, "Ratio: " + ratio);
                set.setDimensionRatio(textureView.getId(), String.valueOf(ratio));
            }
        }

        set.applyTo((ConstraintLayout) textureView.getParent());
    }

    private void updateStudentsUI() {
        List<RemoteLearningParticipant> stds = students.stream().filter(it -> !it.isModerator()).collect(Collectors.toList());
        for (int i = 0; i < stds.size(); i++) {
            viewModel.detachParticipantFromView(stds.get(i).getParticipantId());
            if (stds.get(i).isVideo()) {
                studentTextureViews.get(i).setVisibility(View.VISIBLE);
                attachStudentStream(stds.get(i), studentTextureViews.get(i));
                setStudentTextureViewConstraints(
                        stds.get(i).getWidth(),
                        stds.get(i).getHeight(),
                        studentTextureViews.get(i)
                );
            }
        }
    }

    private void attachStudentStream(RemoteLearningParticipant student, TextureView textureView) {
        viewModel.attachParticipantToView(student.getParticipantId(), textureView);
    }

    private void configureLandscapeView() {
        binding.scrollView.setVisibility(View.GONE);
        setModeratorTextureViewConstraints();
    }

    private void configurePortraitView() {
        binding.scrollView.setVisibility(View.VISIBLE);
        setModeratorTextureViewConstraints();
    }
}

File: sample-app-kotlin/app/src/main/java/com/bluejeans/android/sdksample/isc/usecases/RemoteLearningFragment.kt
package com.bluejeans.android.sdksample.isc.usecases

import android.content.res.Configuration
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.TextureView
import android.view.View
import android.view.ViewGroup
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.constraintlayout.widget.ConstraintSet
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import com.bjnclientcore.media.individualstream.StreamPriority
import com.bjnclientcore.media.individualstream.StreamQuality
import com.bjnclientcore.media.individualstream.VideoStreamConfiguration
import com.bjnclientcore.media.individualstream.VideoStreamConfigurationResult
import com.bjnclientcore.utils.extensions.getInitials
import com.bluejeans.android.sdksample.databinding.FragmentRemoteLearningBinding

class RemoteLearningFragment : Fragment() {

    private var _binding: FragmentRemoteLearningBinding? = null
    private val binding get() = _binding!!

    private lateinit var viewModel: RemoteLearningViewModel

    private var moderator: RemoteLearningParticipant? = null
    private val students = mutableListOf<RemoteLearningParticipant>()
    private val studentTextureViews = mutableListOf<TextureView>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        _binding = FragmentRemoteLearningBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        viewModel = ViewModelProvider(this)[RemoteLearningViewModel::class.java]

        studentTextureViews.addAll(
            listOf(
                binding.studentOne.participantTextureView,
                binding.studentTwo.participantTextureView,
                binding.studentThree.participantTextureView,
                binding.studentFour.participantTextureView,
                binding.studentFive.participantTextureView,
                binding.studentSix.participantTextureView
            )
        )

        viewModel.participantsObservable.observe(viewLifecycleOwner) { studentsList ->
            studentTextureViews.forEach {
                it.visibility = View.GONE
            }
            moderator = null
            students.clear()
            students.addAll(studentsList)

            val configurationList = mutableListOf<VideoStreamConfiguration>()
            for (remoteLearningParticipant in students.distinctBy { it.participantId }) {
                if (remoteLearningParticipant.isModerator) {
                    moderator = remoteLearningParticipant
                    configurationList.add(
                        VideoStreamConfiguration(
                            remoteLearningParticipant.participantId,
                            StreamQuality.R720p_30fps,
                            StreamPriority.High
                        )
                    )
                } else {
                    configurationList.add(
                        VideoStreamConfiguration(
                            remoteLearningParticipant.participantId,
                            StreamQuality.R90p_15fps,
                            StreamPriority.Low
                        )
                    )
                }
            }

            when (val result = viewModel.setVideoConfiguration(configurationList)) {
                is VideoStreamConfigurationResult.Success -> {
                    updateModeratorUI()
                    updateStudentsUI()
                }
                else -> Log.e(TAG, "Failed to set stream configuration: $result")
            }
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        moderator?.let { viewModel.detachParticipantFromView(it.participantId) }
        students.forEach {
            viewModel.detachParticipantFromView(it.participantId)
        }
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        if (newConfig.orientation == Configuration.ORIENTATION_PORTRAIT) {
            configurePortraitView()
        } else configureLandscapeView()
    }

    private fun updateModeratorUI() {
        moderator?.let {
            binding.layoutMainStage.tvParticipantName.text = it.name

            if (!it.isVideo) {
                Log.i(TAG, "Detaching moderator video")
                viewModel.detachParticipantFromView(it.participantId)

                binding.layoutMainStage.participantTextureView.visibility = View.GONE
                binding.layoutMainStage.cnhAudioOnly.visibility = View.VISIBLE
                binding.layoutMainStage.cnhAudioOnly.setText(it.name.getInitials())
            } else {
                binding.layoutMainStage.participantTextureView.visibility = View.VISIBLE
                binding.layoutMainStage.cnhAudioOnly.visibility = View.GONE
                viewModel.attachParticipantToView(
                    it.participantId,
                    binding.layoutMainStage.participantTextureView
                )
                setModeratorTextureViewConstraints()
            }
        } ?: kotlin.run {
            binding.layoutMainStage.tvParticipantName.text = ""
            binding.layoutMainStage.cnhAudioOnly.visibility = View.GONE
        }
    }

    private fun updateStudentsUI() {
        students.filter { !it.isModerator }.forEachIndexed { index, it ->
            viewModel.detachParticipantFromView(it.participantId)
            if (it.isVideo) {
                studentTextureViews[index].visibility = View.VISIBLE
                attachStudentStream(it, studentTextureViews[index])
                if (it.width != null && it.height != null) {
                    setStudentTextureViewConstraints(
                        it.width,
                        it.height,
                        studentTextureViews[index]
                    )
                }
            }
        }
    }

    private fun attachStudentStream(student: RemoteLearningParticipant, textureView: TextureView) {
        viewModel.attachParticipantToView(student.participantId, textureView)
    }

    private fun setModeratorTextureViewConstraints() {
        val set = ConstraintSet()
        val textureView = binding.layoutMainStage.participantTextureView
        set.clone(binding.layoutMainStage.root)
        set.connect(
            textureView.id,
            ConstraintSet.RIGHT,
            ConstraintSet.PARENT_ID,
            ConstraintSet.RIGHT
        )
        set.connect(textureView.id, ConstraintSet.LEFT, ConstraintSet.PARENT_ID, ConstraintSet.LEFT)
        set.connect(
            textureView.id,
            ConstraintSet.BOTTOM,
            ConstraintSet.PARENT_ID,
            ConstraintSet.BOTTOM
        )
        set.connect(textureView.id, ConstraintSet.TOP, ConstraintSet.PARENT_ID, ConstraintSet.TOP)

        moderator?.let {
            if (it.width != null && it.height != null) {
                val ratio = (it.width.toFloat() / it.height)
                Log.i(TAG, "Ratio: $ratio")
                set.setDimensionRatio(textureView.id, ratio.toString())
            }
        }

        set.applyTo(binding.layoutMainStage.root)
    }

    private fun configureLandscapeView() {
        binding.scrollView.visibility = View.GONE
        setModeratorTextureViewConstraints()
    }

    private fun configurePortraitView() {
        binding.scrollView.visibility = View.VISIBLE
        setModeratorTextureViewConstraints()
    }

    private fun setStudentTextureViewConstraints(
        videoWidth: Int,
        videoHeight: Int,
        textureView: TextureView
    ) {
        val set = ConstraintSet()
        set.clone(binding.layoutMainStage.root)
        set.connect(
            textureView.id,
            ConstraintSet.RIGHT,
            ConstraintSet.PARENT_ID,
            ConstraintSet.RIGHT
        )
        set.connect(textureView.id, ConstraintSet.LEFT, ConstraintSet.PARENT_ID, ConstraintSet.LEFT)
        set.connect(
            textureView.id,
            ConstraintSet.BOTTOM,
            ConstraintSet.PARENT_ID,
            ConstraintSet.BOTTOM
        )
        set.connect(textureView.id, ConstraintSet.TOP, ConstraintSet.PARENT_ID, ConstraintSet.TOP)
        if (moderator != null) {
            if (videoWidth > 0 && videoHeight > 0) {
                val ratio = videoWidth.toFloat() / videoHeight
                Log.i(TAG, "Ratio: $ratio")
                set.setDimensionRatio(textureView.id, ratio.toString())
            }
        }
        set.applyTo(textureView.parent as ConstraintLayout)
    }

    companion object {
        private val TAG = "RemoteLearningFragment"
    }
}